### ISSUE TITLE
npu(cascade): single-launch cascade perf tuning (E1/E3 D-launch) — beat P1+P2 4.65 ms for decode-default MoE

### ISSUE THREAD

**Issue by bong-water-water-bong** [1bit-MONSTER/1bit-MONSTER#2113]
Follow-up to the #2078 resolution (commit e3abace8, corr 0.578 -> 0.999346 silicon-verified).

## Status
The correctness-first single-launch fused GU->SiLU->D cascade launch is ~12.3 ms (12.4-14.9 ms incl. per-token pack/fold). The tuned production split (P1 GU+silu->h2 + P2 D) is **4.65 ms** per the M1 profile, so the cascade is NOT yet a decode win — correctness landed first.

## Remaining work
- E1/E3 D-launch tuning on the cascade path (the commit msg names this as the remaining decode-win work).
- Once cascade single-launch < split launch, wire it as the default MoE FFN executor in the decode loop (NPU_CASCADE_DECODE currently proves the layer-1 FFN integration only) and re-run the end-to-end decode perf + logits gate.
- Confirm the fold-element AB stream (+1 fold element per (col,cg)) doesn't regress the batch/all-ones path used by fused_ab_probe.

Perf baseline for reference (M1 profile, 2-token decode): attn 0.71 ms/layer, moe 38.69 ms/layer; [CAS] single-launch 12.4 ms incl. pack.

**Comment by bong-water-water-bong** [comment]
**Disposition (triage, 2026-09-05):** perf task within the cascade M3 scope (single-launch E1/E3 D-launch vs the P1+P2 4.65 ms baseline). **Needs:** silicon benchmark on strixhalo after the M3 whole-layer kernel work (#2080) provides the launch path. Keep open.

**Comment by bong-water-water-bong** [comment]
## The cascade has a ~5 ms fixed launch floor (architectural to the iron/rt.sequence flow)

Measured fixed-cost isolation (no_gu artifacts, batch=4 fills, sequential, deterministic):
```
no_gu K=512 N_D=512 (1/16 the D MACs, 8 b8/col):   5.08 ms
no_gu K=2048 N_D=512 (full K, 32 b8/col):           4.95 ms
no_gu K=2048 N_D=2048 ROWS=4 (4x the cols):         7.27 ms
full nd2048 ROWS=4 (GU+D):                          ~11.8 ms
production P2 (same D, old-API n1_core_fused_gu_silu_d_p2): 1.49 ms
```
Cutting the D MACs 16x and the fills 4x changes the launch time by ~0.1 ms — the ~5 ms is a FIXED per-launch floor. The production kernels (old aie.dialects API) launch the same work in 1.49 ms total. Conclusion: the iron/rt.sequence + dynamic-objFifos launch has a ~5 ms floor that makes the iron-built cascade structurally unable to beat the old-API production kernels (~1.5-3 ms) for decode — independent of fill batching (done, marginal) or D-compute tuning (the floor dominates).

Implication for #2113's goal (beat P1+P2 4.65 ms for decode): the iron-flow cascade cannot win on decode perf. Options: (a) port the cascade's now-validated numerics (per-column fold + full-precision silu + per-column D, corr 0.999346) onto the OLD-API single-launch design (n1_core_fused_gu_silu_d_p2.py / the #1759 fused) if the single-launch-per-layer is wanted for decode; or (b) treat the cascade as the numerics-validation substrate and keep the production old-API kernels for decode (the #2080 whole-layer direction). The ~5 ms floor should be pinned before any further cascade-perf investment.

**Comment by bong-water-water-bong** [comment]
## Item-1 (old-API port) measured — the decode-level port is UNNECESSARY: the old-API single-launch already wins

The port question (#2113 option a) is answered with data, no code needed:

**The production old-API single-launch fused (#1759, `n1_core_fused_gu_silu_d.py` → `final_i8_MOE_FUSED_zaya`, unchanged since Aug 23) is already the decode's fast+correct single-launch:**

Clean sequential N_GEN=2 runs on zaya1-8b (strixhalo, no contention):
```
path            ms/token   moe wall/layer   kernels
single #1759    97.0       6.8 ms           ONE launch (pre 3.4 + kernel+dequant ~3.4)
split P1+P2     157.4      12.7 ms          P1 3.09 + host-silu + h2 sync + P2 1.64
```
- L1 corr 0.998469 vs CPU (x86-validated contract 0.9993-0.9996), deterministic across 3 silicon runs (identical corr/tokens/logits).
- It is the decode's DEFAULT (`FUSED_SINGLE` when `NPU_FUSED=1`); the split is the #1775 nondeterminism fallback.

**Consequences:**
1. The iron cascade (11.8 ms/layer go()) is 1.75x SLOWER than the existing production single at decode level — the iron flow's ~5 ms floor makes it strictly dominated. The cascade's decode-FFN ambition is closed; it stands as the numerics substrate + the full-depth validation harness.
2. The "port" payload (per-column fold scales + q22 silu_pair + multi-row ROWS=4 geometry, corr 0.9993) only has value as a component for the #1776/#2080 WHOLE-LAYER kernel, where a multi-row single-launch MoE stage in the production object-fifo mechanics is needed. If/when the whole-layer work starts, the base = n1_core_fused_gu_silu_d.py's machinery + the cascade's validated fold math — the iron generator retires.
3. The decode-level per-layer fidelity question (0.9985 vs the 0.9999 needed for 20-layer argmax stability) applies EQUALLY to the production single — the fold-sat/qn_s<1 regime work (item 2) should target the production single's quant path, not the iron cascade.

Recommendation: keep the production single-launch as the decode path; file the whole-layer single-launch MoE component as part of #2080's build-out; do NOT port the iron cascade into a production kernel.

**Comment by bong-water-water-bong** [comment]
## #2113 perf re-measurement (2026-09-08, strixhalo, branch npu/cascade-e2b-preserve @1f87ca64)

zaya1-8b.q4nx, 8-token decode, same harness:

| config | total | per-layer MoE |
|---|---|---|
| production (default split P1+P2 + router/amax) | 156–160 ms/tok | **~6.56 ms** (M1prof moe 131.1/10) |
| production + cascade sweep (NPU_CASCADE_SWEEP, compare mode) | 244.8 ms/tok | +4.27 ms/layer = cascade full path (launch + packB + compare) |
| production + cascade in-flow (NPU_CASCADE_DECODE=1) | 251.3 ms/tok | +4.53 ms/layer |

The cascade single-launch full path ≈ **4.3–4.5 ms/layer vs the P1+P2 split bar 4.65 ms** (launch-only) and vs ~6.56 ms/layer production total — the batching commits (88020e0c, 5527e322) appear to have closed the gap. Caveat: the 4.3–4.5 figure includes host-side packB + compare (delta-based, not instrumented); a clean sw.go()-only timing needs a 3-line chrono in the hook + rebuild. Recommend the instrumented re-measure before the "default MoE executor" switch.

**Comment by bong-water-water-bong** [comment]
**CORRECTION to my earlier comment** — the instrumented measurement (commit 6caea991 on cascade-sweep-run, sw.go() timed directly) shows the cascade single-launch is **8.95–11.20 ms/layer (mean 9.76, n=80)** vs the production split total ~6.56 ms/layer and the P1+P2 bar 4.65 ms. The earlier '~4.3-4.5 ms, gap appears closed' was a delta-based artifact (device contention between the sweep and the production launches inflated the production side). **The cascade is still ~1.5x slower than production — #2113 remains open; the E1/E3 D-launch tuning is the work.**

**Comment by bong-water-water-bong** [comment]
_Diagnosis + fix plan — issue campaign (goal `mtubusx1-swek0b`, 2026-09-09). Full inventory: `docs/issue-campaign/FIX-LIST.md`._

**Diagnosis: correctness landed, 2.6× D-phase gap.** Cascade is corr-verified (0.999346) but 12.3 ms vs split 4.65 ms. The D leg = matmul_i8_i32_wide_k8 + cascade_reduce_{first,mid,last}_i32_wide (8 partial sums) + per-token pack/fold; the split's P2 D is a single direct GEMM. Extra ~7.7 ms = wide-N reduce + fold.

**Note:** cascade probes/perf commits live on branch `cascade-sweep-run` (88020e0c, 5527e322, 504160de), not goal/flm-parity-zaya-perf.

**Fix plan:** (1) profile [CASC] go_ms (zaya_decode.cpp:1987) fill/await vs compute vs cascade_reduce; (2) continue batching D-side BD fills; check if the reduce chain serializes — try ROWS>1 memtile fanout or wide-k8 reduce overlap; (3) wire as default MoE FFN only after <4.65 ms, then fold-element AB regression vs fused_ab_probe (bad=0).

—_posted by the issue campaign coordinator_

**Comment by bong-water-water-bong** [comment]
## Status: the stated goal is closed by this issue's own measurements — the last comment proposes work against it anyway

Idle since 2026-09-09, but the substantive conclusion landed on 2026-09-06 and the issue is still titled as an open perf gap.

### The goal is concluded unachievable, with a structural reason

The title's target is *"beat P1+P2 4.65 ms for decode"*. The 09-06T01:51 comment answers it:

> Measured fixed-cost isolation… **Cutting the D MACs 16x and the fills 4x changes the launch time by ~0.1 ms — the ~5 ms is a FIXED per-launch floor.** The production kernels (old `aie.dialects` API) launch the same work in 1.49 ms total. Conclusion: the iron/`rt.sequence` + dynamic-objFifos launch has a ~5 ms floor that makes the iron-built cascade **structurally unable to beat** the old-API production kernels (~1.5–3 ms) for decode — independent of fill batching (done, marginal) or D-compute tuning (the floor dominates).

The measurements behind that are the right shape: `no_gu K=512 N_D=512` at 1/16th the D MACs still costs 5.08 ms against 4.95 ms at full K, so the cost is demonstrably not in the work.

### And the alternative was measured, not assumed

The 09-06T07:29 comment answers the issue's own "remaining work" item (the option-(a) old-API port) with data:

```
path            ms/token   moe wall/layer   kernels
single #1759    97.0       6.8 ms           ONE launch
split P1+P2     157.4      12.7 ms          P1 + host-silu + h2 sync + P2
```

with L1 corr 0.998469 vs CPU, deterministic across three silicon runs. It goes on to state the outcome plainly:

> **The cascade's decode-FFN ambition is closed**; it stands as the numerics substrate + the full-depth validation harness.

So all three of the issue's remaining-work items are discharged: E1/E3 D-launch tuning cannot help (floor dominates), the port is unnecessary (already delivered), and the "wire as default once <4.65 ms" precondition is unreachable by construction.

### The contradiction worth flagging

The **last** comment (09-09T17:01) proposes a fix plan — *"profile `[CASC] go_ms`… continue batching D-side BD fills… wire as default MoE FFN only after <4.65 ms"*. That is three days **after** the floor measurement, and step 3 re-asserts the exact precondition the 09-06 comment showed cannot be met. It reads as a plan written from the issue body without the conclusion in hand — the same failure the tracker keeps producing: **a conclusion recorded in a comment, and a body still describing an open task.**

I am not proposing to re-open the technical question. The 09-06 measurements are self-consistent and the floor explanation is testable and was tested.

### Recommendation

Close, or retitle to the purpose the thread actually settled on. Two things survive and neither is "beat 4.65 ms":

1. **The cascade as numerics substrate / full-depth validation harness** — its real remaining value, and the place the `fix/2114-cascade-qn-guard` work (#2114) belongs.
2. **The port payload** (per-column fold scales + q22 silu_pair + multi-row `ROWS=4` geometry) — explicitly scoped by the 09-06 comment as a component for the **#1776/#2080 whole-layer kernel**, not for decode.

If it stays open, the useful edit is a retitle to (1)/(2), because as written it invites exactly the re-litigation the 09-09 comment represents.

**Scope:** read-only — this thread only. Nothing run on the NPU, no files changed, nothing pushed. I did not verify the perf numbers independently; they are quoted as the issue's own record, and my contribution is limited to noting that the issue's *state* does not match its recorded conclusion.

**Comment by bong-water-water-bong** [comment]
Re-checked as of 2026-09-22: same as #2114 — no commits since 2026-09-17 touch the cascade path. The cascade single-launch (12.3-14.9 ms) is still slower than the tuned split (4.65 ms), so it's still not a decode win; E1/E3 D-launch tuning hasn't started.

Status unchanged: PARKED, gated on #2114's correctness sweep landing first (per this repo's own triage ordering).
